### PR TITLE
do s2i git cloning up front, not in s2i itself

### PR DESCRIPTION
**Pull request by bparees**
depends on https://github.com/openshift/source-to-image/pull/660

### CODE REVIEW

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
[testextended][extended:core(builds|image_ecosystem)]

**Comment by bparees** [comment]
@csrwng ptal.  This doesn't totally eliminate the "downloader" override because the concept of having a "downloader" is pretty intrinsic in s2i, but at least now the downloader we pass is basically a no-op and all the actual source cloning/setup is done up front in a similar way to how its done for docker builds.

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/12236
(but also some legit looking issues)

**Comment by bparees** [comment]
@csrwng ok, *now* ptal

File: pkg/build/builder/sti.go [hunk 1]
```diff
@@ -134,20 +113,34 @@ func (s *S2IBuilder) Build() error {
...
+			sourceInfo.ContextDir = s.build.Spec.Source.ContextDir
+		}
+		srcDir = filepath.Join(srcDir, s.build.Spec.Source.ContextDir)
+	}
+	download := &downloader{}
+	if sourceInfo != nil {
+		download.sourceInfo = &sourceInfo.SourceInfo
+		revision := updateBuildRevision(s.build, sourceInfo)
+		if updateErr := retryBuildStatusUpdate(s.build, s.client, revision); updateErr != nil {
+			utilruntime.HandleError(fmt.Errorf("error: An error occured while updating the build status: %v", updateErr))
```
Comment (csrwng) [inline pkg/build/builder/sti.go]: s/error: An //

**Comment by csrwng** [comment]
One really minor comment, otherwise LGTM

File: pkg/build/builder/sti.go (hunk 1 above)
Comment (bparees) [inline pkg/build/builder/sti.go]: fixed there and elsewhere.

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/11016
[test]

**Comment by bparees** [comment]
[merge]

**Pull request merged**